### PR TITLE
[FSDP][1/N]Add device_mesh to FSDPstate

### PR DESCRIPTION
**Pull request by wz337**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102551

This PR creates a device_mesh and share it across all FSDP state. The device_mesh will later be used to test out dtensor state_dict (1d device_mesh).
Approved by: https://github.com/awgu

Add device mesh to fsdp state
skip dist.get_world_size(pg) != dist.get_world_size()
address test_fake_pg.py test failure
fix test_fake_py.py failure

cc @zhaojuanmao @mrshenli @rohan-varma @awgu

### CODE REVIEW

**Review by fegin — APPROVED** [review]
LGTM

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -202,6 +204,26 @@ def _check_flat_params_on_expected_device(state: _FSDPState, module: nn.Module):
...
+) -> Optional[DeviceMesh]:
+    # We are testing 1D DeviceMesh where dist.get_world_size(pg) == dist.get_world_size() for now.
+    # TODO: Address cases when dist.get_world_size(pg) != dist.get_world_size(). This would capture
+    #       what 1D DeviceMesh currently would not work for:
+    #       1) HSDP Hybrid Sharding, 2) 2D FSDP + TP, 3) dist.new_group() cannot be expressed in 1D DeviceMesh.
+    if get_world_size(root_state.process_group) != get_world_size():
+        return None
+    # Temporarily skip DeviceMesh initialization when the backend is set to fake.
+    # TODO: Remove the condition once the issue of "Default PG backend: fake not supporting CUDA!" is resolved.
+    if get_backend() == "fake" or not root_state.compute_device:
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/fsdp/_runtime_utils.py:218]: Ye. You are right. Will remove this todo.

File: torch/distributed/fsdp/_runtime_utils.py [hunk 2]
```diff
@@ -210,6 +212,26 @@ def _check_flat_params_on_expected_device(state: _FSDPState, module: nn.Module):
             )
 
 
+def _init_device_mesh(
+    root_state: _FSDPState,
+) -> Optional[DeviceMesh]:
+    # We are testing 1D DeviceMesh where dist.get_world_size(pg) == dist.get_world_size() for now.
+    # TODO: Address cases when dist.get_world_size(pg) != dist.get_world_size(). This would capture
+    #       what 1D DeviceMesh currently would not work for:
+    #       1) HSDP Hybrid Sharding, 2) 2D FSDP + TP, 3) dist.new_group() cannot be expressed in 1D DeviceMesh.
```
Comment (kumpera) [inline torch/distributed/fsdp/_runtime_utils.py:221]: Point (3) is not true, we can create a mesh with just a few ranks.

File: torch/distributed/fsdp/_runtime_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kumpera) [reply torch/distributed/fsdp/_runtime_utils.py:226]: I think we should still try to figure out why it doesn't work with one of our testing backends.

File: torch/distributed/fsdp/_runtime_utils.py [hunk 3]
```diff
@@ -210,6 +212,26 @@ def _check_flat_params_on_expected_device(state: _FSDPState, module: nn.Module):
...
+    if get_world_size(root_state.process_group) != get_world_size():
+        return None
+    # Temporarily skip DeviceMesh initialization when the backend is set to fake.
+    # TODO: Remove the condition once the issue of "Default PG backend: fake not supporting CUDA!" is resolved.
+    if get_backend() == "fake" or not root_state.compute_device:
+        return None
+    device_type = root_state.compute_device.type
+    mesh_tensor = torch.arange(get_world_size(root_state.process_group))
+    device_mesh = DeviceMesh(device_type, mesh_tensor, _init_process_groups=False)
+    device_mesh._dim_groups = [root_state.process_group]
```
Comment (kumpera) [inline torch/distributed/fsdp/_runtime_utils.py:231]: We should fix the PG<>DM interop as part of this effort. This piece of code is not particularly great.
CC @wanchaol

**Review by kumpera — COMMENTED** [review]
Love it!

**Pull request closed without merging**